File: spire_blue_work/notes/2018/二季度/5月/2018-5-17/LeetCode/leetcode.3/java/solution1/Program.java
import java.util.*;
import java.lang.*;


class Program
{


	public static int LengthOfLongestSubstring(String s) {
		if (s.length() < 2) { return s.length(); }

		int res = 0;


		Map<Character,Integer> arr = new HashMap<Character,Integer>();
		int j=0;
		char c=0;
		for (int i = 0; i < s.length(); i++)
		{
			if (arr.containsKey(s.charAt(i)))
			{
				c=s.charAt(i);
				arr.put(c,arr.get(c)+1);
				//arr.get((char)c)++;
				//arr.get(new Character(s.charAt(i))) = arr.get(s.charAt(i)) + 1;
			}
			else {
				arr.put(s.charAt(i),1);
			
			}


			

				while (arr.get(s.charAt(i)) > 1) {
					if (arr.get(s.charAt(i)) > 1)
					{
						c=s.charAt(j);
						//--arr.get(s.charAt(j));
						arr.put(c,arr.get(c)-1);
						j++;
					}
				}

			res = Math.max(res,i-j+1);

			
		}
		arr.clear();

		return res;

	
	}


   public static void main(String[] args)
	{
	   System.out.println(LengthOfLongestSubstring("aabbbb"));
		System.out.println(LengthOfLongestSubstring("bbbbbb"));

		System.out.println(LengthOfLongestSubstring("pwwkew"));
		System.out.println();


	}
}

File: spire_blue_work/notes/2018/二季度/5月/2018-5-17/LeetCode/leetcode.3/C#/solution1/Program.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

namespace playtheCode
{
    class Program
    {

        /// <summary>
        /// 
        /// </summary>
        /// <param name="s"></param>
        /// <returns></returns>
        public static int LengthOfLongestSubstring(string s) {
            if (s.Length < 2) { return s.Length; }

            int res = 0;

          //  int []arr=new int[s.Length];
            Dictionary<char, int> arr = new Dictionary<char, int>();
            int j=0;
            for (int i = 0; i < s.Length; i++)
            {
                if (arr.ContainsKey(s[i]))
                {
                    arr[s[i]] = arr[s[i]] + 1;
                }
                else {
                    arr.Add(s[i],1);
                
                }


                

                    while (arr[s[i]] > 1) {
                        if (arr[s[i]] > 1)
                        {
                        //arr[s[i]] = arr[s[j++]] - 1;//some error
                            --arr[s[j]];
                            j++;
                        }
                    }

                res = Math.Max(res,i-j+1);

                
            }


            return res;

        
        }


        static void Main(string[] args)
        {
           Console.WriteLine(LengthOfLongestSubstring("aabbbb"));
            Console.WriteLine(LengthOfLongestSubstring("bbbbbb"));

            Console.WriteLine(LengthOfLongestSubstring("pwwkew"));
            Console.ReadKey();


        }
    }
}
